lms/djangoapps/edxnotes/helpers.py

```diff
@@ -0,0 +1,388 @@
+from django.core.exceptions import ImproperlyConfigured
+from django.utils.translation import ugettext as _
+
+from capa.util import sanitize_html
+from student.models import anonymous_id_for_user
+from xmodule.modulestore.django import modulestore
+from xmodule.modulestore.exceptions import ItemNotFoundError
+from util.date_utils import get_default_time_display
+from dateutil.parser import parse as dateutil_parse
+from provider.oauth2.models import AccessToken, Client
+import oauth2_provider.oidc as oidc
+from provider.utils import now
+from opaque_keys.edx.keys import UsageKey
+from .exceptions import EdxNotesParseError, EdxNotesServiceUnavailable
+
+log = logging.getLogger(__name__)
+HIGHLIGHT_TAG = "span"
+HIGHLIGHT_CLASS = "note-highlight"
+
+
+class NoteJSONEncoder(JSONEncoder):
+    """
+    Custom JSON encoder that encode datetime objects to appropriate time strings.
+    """
+    # pylint: disable=method-hidden
+    def default(self, obj):
+        if isinstance(obj, datetime):
+            return get_default_time_display(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def get_id_token(user):
+    """
+    Generates JWT ID-Token, using or creating user's OAuth access token.
+    """
+    try:
+        client = Client.objects.get(name="edx-notes")
+    except Client.DoesNotExist:
+        raise ImproperlyConfigured("OAuth2 Client with name 'edx-notes' is not present in the DB")
+    try:
+        access_token = AccessToken.objects.get(
+            client=client,
+            user=user,
+            expires__gt=now()
+        )
+    except AccessToken.DoesNotExist:
+        access_token = AccessToken(client=client, user=user)
+        access_token.save()
+
+    id_token = oidc.id_token(access_token)
+    secret = id_token.access_token.client.client_secret
+    return id_token.encode(secret)
+
+
+def get_token_url(course_id):
+    """
+    Returns token url for the course.
+    """
+    return reverse("get_token", kwargs={
+        "course_id": unicode(course_id),
+    })
+
+
+def send_request(user, course_id, path="", query_string=None):
+    """
+    Sends a request with appropriate parameters and headers.
+    """
+    url = get_endpoint(path)
+    params = {
+        "user": anonymous_id_for_user(user, None),
+        "course_id": unicode(course_id).encode("utf-8"),
+    }
+
+    if query_string:
+        params.update({
+            "text": query_string,
+            "highlight": True,
+            "highlight_tag": HIGHLIGHT_TAG,
+            "highlight_class": HIGHLIGHT_CLASS,
+        })
+
+    try:
+        response = requests.get(
+            url,
+            headers={
+                "x-annotator-auth-token": get_id_token(user)
+            },
+            params=params
+        )
+    except RequestException:
+        raise EdxNotesServiceUnavailable(_("EdxNotes Service is unavailable. Please try again in a few minutes."))
+
+    return response
+
+
+def get_parent_xblock(xblock):
```

Comment:
I know, but what happens if code accesses the parent after a reparent within the same request. I guess that's more of a problem for @jimabramson's generic solution than for yours, since you know where you are using this method.

Anyway, there need to be tests of the basic cases. What happens when called on a block which hasn't been parented yet? What happens with the boundary between a draft child and a non-draft parent? Are there gotchas in split vs old mongo?